golang/trace/zipin_tracereporter_test.go

```diff
@@ -0,0 +1,187 @@
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"github.com/uber/tchannel/golang"
+	"github.com/uber/tchannel/golang/thrift"
+	gen "github.com/uber/tchannel/golang/trace/thrift/gen-go/tcollector"
+	"github.com/uber/tchannel/golang/trace/thrift/mocks"
+)
+
+func TestZipkinTraceReporterFactory(t *testing.T) {
+	_, err := tchannel.NewChannel("client", &tchannel.ChannelOptions{
+		Logger:               tchannel.SimpleLogger,
+		TraceReporterFactory: ZipkinTraceReporterFactory,
+	})
+
+	assert.NoError(t, err)
+}
+
+func TestBuildZipkinSpan(t *testing.T) {
+	host := &tchannel.Endpoint{
+		Ipv4:        "127.0.0.1",
+		Port:        8888,
+		ServiceName: "test",
+	}
+	span := *tchannel.NewRootSpan()
+	name := "test"
+	annotations := RandomAnnotations()
+	thriftSpan := buildZipkinSpan(span, annotations, nil, name, host)
+
+	expectedSpan := &gen.Span{
+		TraceId: uint64ToBytes(span.TraceID()),
+		Host: &gen.Endpoint{
+			Ipv4:        (int32)(inetAton("127.0.0.1")),
+			Port:        8888,
+			ServiceName: "test",
+		},
+		Name:        name,
+		Id:          uint64ToBytes(span.SpanID()),
+		ParentId:    uint64ToBytes(span.ParentID()),
+		Annotations: buildZipkinAnnotations(annotations),
+		Debug:       false,
+	}
+
+	assert.Equal(t, thriftSpan, expectedSpan, "Span mismatch")
+}
+
+func TestInetAton(t *testing.T) {
+	assert.Equal(t, inetAton("1.2.3.4"), uint32(16909060))
+}
+
+func TestUInt64ToBytes(t *testing.T) {
+	assert.Equal(t, uint64ToBytes(54613478251749257), []byte("\x00\xc2\x06\xabK$\xdf\x89"))
+}
+
+func TestBase64Encode(t *testing.T) {
+	assert.Equal(t, base64Encode(12711515087145684), "AC0pDj1TitQ=")
+}
+
+func RandomAnnotations() []tchannel.Annotation {
+	baseTime := time.Date(2015, 1, 2, 3, 4, 5, 6, time.UTC)
+	return []tchannel.Annotation{
+		{
+			Key:       tchannel.AnnotationKeyClientReceive,
+			Timestamp: baseTime.Add(time.Second),
+		},
+		{
+			Key:       tchannel.AnnotationKeyClientSend,
+			Timestamp: baseTime.Add(2 * time.Second),
+		},
+		{
+			Key:       tchannel.AnnotationKeyServerReceive,
+			Timestamp: baseTime.Add(3 * time.Second),
+		},
+		{
+			Key:       tchannel.AnnotationKeyServerSend,
+			Timestamp: baseTime.Add(4 * time.Second),
+		},
+	}
+}
+
+type testArgs struct {
+	s *mocks.TChanTCollector
+	c tchannel.TraceReporter
+}
+
+func ctxArg() mock.AnythingOfTypeArgument {
+	return mock.AnythingOfType("*tchannel.headerCtx")
+}
+
+func TestSubmit(t *testing.T) {
+	withSetup(t, func(ctx thrift.Context, args testArgs) {
+		host := &tchannel.Endpoint{
+			Ipv4:        "127.0.0.1",
+			Port:        8888,
+			ServiceName: "test",
+		}
+		span := *tchannel.NewRootSpan()
+		name := "test"
+		annotations := RandomAnnotations()
+		thriftSpan := buildZipkinSpan(span, annotations, nil, name, host)
+		thriftSpan.BinaryAnnotations = []*gen.BinaryAnnotation{}
+		ret := &gen.Response{Ok: true}
+		args.s.On("Submit", ctxArg(), thriftSpan).Return(ret, nil)
```

Comment:
nit: probably add a blank line above this to separate the data / setup from the actual call and verification